model/riscv_types.sail

```diff
@@ -410,3 +412,41 @@ function report_invalid_width(f , l, w, k) -> 'a = {
   internal_error(f, l, "Invalid width, " ^ size_mnemonic(w) ^ ", for " ^ k ^
      " with xlen=" ^ dec_str(xlen))
 }
+
+/* Callbacks for state-changing events */
+
+/* Defaults for these functions in riscv_default_callbacks.c */
+
+val mem_write_callback = pure {c: "mem_write_callback"} : forall 'n, 0 < 'n <= max_mem_access . (/* addr */ xlenbits, /* width */ int('n), /* value */ bits(8 * 'n)) -> unit
+val mem_read_callback = pure {c: "mem_read_callback"} : forall 'n, 0 < 'n <= max_mem_access . (/* access type */ string, /* addr */ xlenbits, /* width */ int('n), /* value */ bits(8 * 'n)) -> unit
+val mem_exception_callback = pure {c: "mem_exception_callback"} : forall 'n, 0 <= 'n < xlen . (/* addr */ xlenbits, /* num_of_ExceptionType */ int('n)) -> unit
+val pc_write_callback = pure {c: "pc_write_callback"} : xlenbits -> unit
+val xreg_write_callback = pure {c: "xreg_write_callback"} : (regidx, xlenbits) -> unit
+val freg_write_callback = pure {c: "freg_write_callback"} : (regidx, flenbits) -> unit
+val csr_write_callback = pure {c: "csr_write_callback"} : (csreg, xlenbits) -> unit
+val csr_read_callback = pure {c: "csr_read_callback"} : (csreg, xlenbits) -> unit
+
+/* Implementations of default callbacks for trace printing */
+
+val mem_write_callback_default : forall 'n, 0 < 'n <= max_mem_access . (/* addr */ xlenbits, /* width */ int('n), /* value */ bits(8 * 'n)) -> unit
+function mem_write_callback_default (addr, width, value) = {
+  print_mem("mem[" ^ BitStr(addr) ^ "] <- " ^ BitStr(value))
+}
+
+val mem_read_callback_default : forall 'n, 0 < 'n <= max_mem_access . (/* access type */ string, /* addr */ xlenbits, /* width */ int('n), /* value */ bits(8 * 'n)) -> unit
+function mem_read_callback_default (t, addr, width, value) = {
+  print_mem("mem[" ^ t ^ "," ^ BitStr(addr) ^ "] -> " ^ BitStr(value))
+}
+
+val xreg_write_callback_default : (regidx, xlenbits) -> unit
+function xreg_write_callback_default (reg, value) = {
+  print_reg("x" ^ dec_str(regidx_to_regno(reg)) ^ " <- " ^ BitStr(value))
+}
+
+val freg_write_callback_default : (regidx, flenbits) -> unit
+function freg_write_callback_default (reg, value) = {
+  /* todo: will only print bits; should we print in floating point format? */
+  print_reg("f" ^ dec_str(regidx_to_regno(reg)) ^ " <- " ^ BitStr(value))
+}
```

Comment:
Now that the OCaml simulator is gone, I wonder if we should just move these default callbacks to C? Doing the sail -> C -> sail roundtrip seems a bit unnecessary?

Reply:
I agree, it's done.